core/src/main/java/io/grpc/Metadata.java

```diff
@@ -235,7 +305,25 @@ public boolean containsKey(Key<?> key) {
   public <T> void put(Key<T> key, T value) {
     Preconditions.checkNotNull(key, "key");
     Preconditions.checkNotNull(value, "value");
-    storeAdd(key.name, new MetadataEntry(key, value));
+    if (len() == 0 || len() == cap()) {
+      int desiredCap = len() + (len() >>> 1) + 4;
+      expand(desiredCap & ~1); // floor to a multiple of 2
+    }
+    name(size, key.asciiName());
+    value(size, key.toBytes(value));
+    typedValues[size] = value;
+    size++;
+  }
+
+  private void expand(int newCapacity) {
+    byte[][] newNamesAndValues = new byte[newCapacity][];
+    Object[] newTypedValues = new Object[newCapacity];
+    if (size != 0) {
```

Comment:
There are a lot of places where we return `new Metadata()` that will never be mutated or really used.  The size check saves a null check.